ui/src/layouts/auth.rs

```diff
@@ -0,0 +1,129 @@
+use dioxus::prelude::*;
+use dioxus_router::use_router;
+use futures::StreamExt;
+use kit::{
+    elements::{
+        button::Button,
+        input::{Input, Options, Validation},
+    },
+    icons::Icon,
+};
+use tokio::sync::oneshot;
+
+use crate::{
+    warp_runner::{commands::TesseractCmd, WarpCmd},
+    CHAT_ROUTE, WARP_CMD_CH,
+};
+
+#[allow(non_snake_case)]
+pub fn AuthLayout(cx: Scope) -> Element {
+    let warp_cmd_tx = WARP_CMD_CH.0.clone();
+    let router = use_router(cx);
+    let username = use_state(cx, String::new);
+    //let error = use_state(cx, String::new);
+    let username_valid = use_state(cx, || false);
+    let pin = use_state(cx, String::new);
+    let pin_valid = use_state(cx, || false);
+
+    let username_validation = Validation {
+        // The input should have a maximum length of 32
+        max_length: Some(32),
+        // The input should have a minimum length of 4
+        min_length: Some(4),
+        // The input should only contain alphanumeric characters
+        alpha_numeric_only: true,
+        // The input should not contain any whitespace
+        no_whitespace: true,
+    };
+
+    // Set up validation options for the input field
+    let pin_validation = Validation {
+        // The input should have a maximum length of 32
+        max_length: Some(32),
+        // The input should have a minimum length of 4
+        min_length: Some(4),
+        // The input should only contain alphanumeric characters
+        alpha_numeric_only: false,
+        // The input should not contain any whitespace
+        no_whitespace: true,
+    };
+
+    let ch = use_coroutine(cx, |mut rx: UnboundedReceiver<(String, String)>| {
+        to_owned![warp_cmd_tx, router];
+        async move {
+            while let Some((username, pin)) = rx.next().await {
+                //println!("auth got input");
+                let (tx, rx) = oneshot::channel::<Result<(), warp::error::Error>>();
+
+                warp_cmd_tx
+                    .send(WarpCmd::Tesseract(TesseractCmd::CreateIdentity {
+                        username: username.into(),
```

Comment:
Due to username being a `String` there is no need to convert it to `String`
```suggestion
                        username,
```